File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/singleton/abstract_factory/solution/Garden.java
// Garden.java

public abstract class Garden
{
// Singleton:

	protected Garden(Lettuce aLettucePrototype)
	{
		setLettucePrototype(aLettucePrototype);
	}

	public static Garden getInstance()
	{
		if (theInstance == null)
		{
			if (true)
			{
				theInstance = new ConventionalGarden(new ConventionalLettuce(1, 2));
			}
			else
			{
				theInstance = new OrganicGarden(new OrganicLettuce(1, 2));
			}
		}

		return theInstance;
	}

	private static Garden theInstance = null;

// non-Singleton:

	protected void setLettucePrototype(Lettuce aLettucePrototype)
	{
		if (aLettucePrototype == null)
		{
			throw new DpException("void setLettucePrototype(Lettuce aLettucePrototype) -- aLettucePrototype == null.");
		}

		myLettucePrototype = aLettucePrototype.clone();
	}

	public Lettuce makeLettuce()
	{
		return myLettucePrototype.clone();
	}

	public abstract Tomato makeTomato();

	private Lettuce myLettucePrototype;

	private static class ConventionalGarden extends Garden
	{
		public ConventionalGarden(ConventionalLettuce aLettucePrototype)
		{
			super(aLettucePrototype);
		}

		public void setLettucePrototype(ConventionalLettuce aLettucePrototype)
		{
			super.setLettucePrototype(aLettucePrototype);
		}

		public Tomato makeTomato()
		{
			return new ConventionalTomato(1, 2);
		}
	}

	private static class OrganicGarden extends Garden
	{
		public OrganicGarden(OrganicLettuce aLettucePrototype)
		{
			super(aLettucePrototype);
		}

		public void setLettucePrototype(OrganicLettuce aLettucePrototype)
		{
			super.setLettucePrototype(aLettucePrototype);
		}

		public Tomato makeTomato()
		{
			return new OrganicTomato(1, 2);
		}
	}
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/singleton/abstract_factory/solution/garden.cpp
// garden.cpp

#include "conventional_garden.h"
#include "conventional_lettuce.h"
#include "dp_exception.h"
#include "lettuce.h"
#include "organic_garden.h"
#include "organic_lettuce.h"
#include "garden.h"
#include "tomato.h"

// Singleton:

garden * garden::the_instance(nullptr);

garden::garden(lettuce * a_lettuce_prototype)
{
	set_lettuce_prototype(a_lettuce_prototype);
}

garden::~garden()
{
}

garden & garden::get_instance()
{
	if (the_instance == nullptr)
	{
		if (true)
		{
			the_instance = new conventional_garden(new conventional_lettuce(1, 2));
		}
		else
		{
			the_instance = new organic_garden(new organic_lettuce(1, 2));
		}
	}
	return *the_instance;
}

// non-Singleton:

void garden::set_lettuce_prototype(lettuce * a_lettuce_prototype)
{
	if (a_lettuce_prototype == nullptr)
	{
		throw dp_exception("void garden::set_lettuce_prototype(lettuce * a_lettuce_prototype) -- a_lettuce_prototype == nullptr.");
	}

	my_lettuce_prototype = a_lettuce_prototype->clone();
}

lettuce * garden::make_lettuce() const
{
	return my_lettuce_prototype->clone();
}